README.md
<p align="center">
    <a href="https://andergoig.github.io/SwiftInstagram">
        <img src="https://rawgit.com/AnderGoig/SwiftInstagram/master/.assets/SwiftInstagram-Logo.png" alt="SwiftInstagram Logo" width="800" height="248">
    </a>
</p>

<p align="center">
    <a href="https://cocoapods.org/pods/SwiftInstagram">
        <img src="https://img.shields.io/cocoapods/p/SwiftInstagram.svg"
             alt="Platforms">
    </a>
    <a href="https://developer.apple.com/swift">
        <img src="https://img.shields.io/badge/Swift-4.0-orange.svg?style=flat"
             alt="Swift 4.0">
    </a>
    <a href="https://travis-ci.org/AnderGoig/SwiftInstagram/branches">
        <img src="https://img.shields.io/travis/AnderGoig/SwiftInstagram/master.svg"
             alt="Travis">
    </a>
    <a href="https://codebeat.co/projects/github-com-andergoig-swiftinstagram-master">
        <img src="https://codebeat.co/badges/e5274943-8feb-4e51-a937-43552205f62e"
             alt="Codebeat">
    </a>
    <a href="https://andergoig.github.io/SwiftInstagram">
        <img src="https://rawgit.com/AnderGoig/SwiftInstagram/master/docs/badge.svg"
             alt="Documentation">
    </a>
    <a href="https://raw.githubusercontent.com/AnderGoig/SwiftInstagram/master/LICENSE">
        <img src="https://img.shields.io/cocoapods/l/SwiftInstagram.svg"
             alt="License">
    </a>
</p>

<p align="center">
    <a href="https://cocoapods.org/pods/SwiftInstagram">
        <img src="https://img.shields.io/cocoapods/v/SwiftInstagram.svg"
             alt="CocoaPods compatible">
    </a>
    <a href="https://github.com/Carthage/Carthage">
        <img src="https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat"
             alt="Carthage compatible">
    </a>
    <a href="https://github.com/apple/swift-package-manager">
        <img src="https://img.shields.io/badge/Swift%20Package%20Manager-compatible-brightgreen.svg"
             alt="Swift Package Manager">
    </a>
</p>

---

**SwiftInstagram** is a **wrapper** for the **Instagram API** written in **Swift**. It allows you to authenticate users and request data from Instagram effortlessly.

## Features

- [x] Effortless and straightforward use.
- [x] Up-to-date with the latest [Instagram API](https://www.instagram.com/developer/) features.
- [x] Secure access token storing, thanks to [keychain-swift](https://github.com/evgenyneu/keychain-swift).
- [x] Client-Side authentication.
- [x] 100% documented code.
- [x] Use of Swift 4 Decodable.
- [x] Up-to-date with the latest Apple technologies (Swift 4, Xcode 9, etc).

### [Upcoming features](https://github.com/AnderGoig/SwiftInstagram/projects/1)

## Requirements

- iOS 9.0+
- Xcode 9.0+

## Usage

I've prepared a [wiki page](https://github.com/AnderGoig/SwiftInstagram/wiki) where you can find tons of useful things.

* [Installation Guide](https://github.com/AnderGoig/SwiftInstagram/wiki/Installation-Guide) - Follow it to integrate SwiftInstagram into your project.
* [Authentication](https://github.com/AnderGoig/SwiftInstagram/wiki/Authentication) - To learn how to authenticate your Instagram users.
* [Data Retrieval](https://github.com/AnderGoig/SwiftInstagram/wiki/Data-Retrieval) - To start requesting some data from Instagram.
* [API Reference](https://andergoig.github.io/SwiftInstagram/Classes/Instagram.html) - For a more detailed code reference.

## Contributing

If you have feature requests or bug reports, feel free to help out by sending pull requests or by [creating new issues](https://github.com/AnderGoig/SwiftInstagram/issues/new). Please take a moment to
review the [CONTRIBUTING](.github/CONTRIBUTING.md) guidelines.

## Credits

SwiftInstagram is brought to you by [Ander Goig](https://github.com/AnderGoig) and [contributors to the project](https://github.com/AnderGoig/SwiftInstagram/contributors). If you're using SwiftInstagram in your project, attribution would be very appreciated.

#### Companion libraries

SwiftInstagram uses [keychain-swift](https://github.com/evgenyneu/keychain-swift) by [@evgenyneu](https://github.com/evgenyneu) to safely store the access token retrieved by the authentication process.

## License

SwiftInstagram is released under the MIT license. See [LICENSE](LICENSE) for details.

## Author

Ander Goig, [goig.ander@gmail.com](mailto:goig.ander@gmail.com)

[https://github.com/AnderGoig/SwiftInstagram](https://github.com/AnderGoig/SwiftInstagram)


CHANGELOG.md
# Changelog

## [Unreleased]

## [1.1.1] (2017-04-06)
[Full Changelog](https://github.com/AnderGoig/SwiftInstagram/compare/v1.1.0...v1.1.1)
### Added
- Compatibility with Swift 4.1.
### Fixed
- Fix bug with media objects JSON decoding (#26).

## [1.1.0] (2017-01-21)
[Full Changelog](https://github.com/AnderGoig/SwiftInstagram/compare/v1.0.6...v1.1.0)
### Added
- You can now use your own authentication method ([Read more](https://github.com/AnderGoig/SwiftInstagram/wiki/Authentication)).
- New success handlers on:
    - `createComment()`
    - `deleteComment()`
    - `like(media: ...)`
    - `unlike(media: ...)`
- Search media and location by latitude and longitude:
    - `func searchMedia(latitude: Double? = nil, longitude: Double? = nil, ...)`
    - `func searchLocation(latitude: Double? = nil, longitude: Double? = nil, ...)`
- New option to get all the permission scopes on login:
    - `login(..., withScopes: [.all], ...)`
### Changed
- `retrieveAccessToken()` method is now public (#15).
- `storeAccessToken()` is also public (#17).
### Fixed
- Problem with all the HTTP POST requests (e.g. #20).

## [1.0.6] (2017-11-03)
[Full Changelog](https://github.com/AnderGoig/SwiftInstagram/compare/v1.0.5...v1.0.6)
### Added
- JSON parser improvements.
- General source code improvements.
### Changed
- `searchLocation()` method attributes `lat` and `lng` have been replaced by `coordinates`.
- `isSessionValid()` method has been renamed to `isAuthenticated` (property).
- Change key for keychain (#11).
- Better handling of #7 bug.

## [1.0.5] (2017-10-18)
[Full Changelog](https://github.com/AnderGoig/SwiftInstagram/compare/v1.0.4...v1.0.5)
### Changed
- The `Success` and `Failure` handlers, for all the API endpoints, are no longer nullable.
### Fixed
- Fixed the bug parsing `InstagramMedia` objects with a `location` property (#7).

## [1.0.4] (2017-10-10)
[Full Changelog](https://github.com/AnderGoig/SwiftInstagram/compare/v1.0.3...v1.0.4)
### Added
- 100% documented code.
### Changed
- Now, the redirection URI must be defined in the Info.plist file. See [wiki](https://github.com/AnderGoig/SwiftInstagram/wiki/Authentication).
- The `createdTime` attribute of the classes `InstagramComment` and `InstagramMedia` has been renamed to `createdDate` and returns an object of type `Date`.
- Improves [Codebeat](https://codebeat.co/projects/github-com-andergoig-swiftinstagram-master) GBA.
### Fixed
- Fixed the type of the returned errors.

## [1.0.3] (2017-10-08)
[Full Changelog](https://github.com/AnderGoig/SwiftInstagram/compare/v1.0.2...v1.0.3)
### Changed
- Improved documentation.
- `InstagramAuthScope` renamed to `InstagramScope`.

## [1.0.2] (2017-10-07)
[Full Changelog](https://github.com/AnderGoig/SwiftInstagram/compare/v1.0.1...v1.0.2)
### Added
- `InstagramLoginViewController` hides automatically when concludes.
- `@discardableResult` to `logout()` method.
### Changed
- Better code organization, especially for networking.
### Fixed
- Authorization process errors are now correctly captured.
- Travis CI build error.

## [1.0.1] (2017-10-03)
[Full Changelog](https://github.com/AnderGoig/SwiftInstagram/compare/v1.0.0...v1.0.1)
### Added
- Use of `DispatchQueue` for parsing JSON and calling callback functions.
### Changed
- The source code is now a little simpler thanks to the use of `typealias`.
- The structure of the functions has changed, there are two callbacks now (`success` and `failure`) instead of just one.
### Fixed
- Login page `WKWebView` now fits correctly when rotating device.

## [1.0.0] (2017-09-30)
### Added
- Initial release.

[Unreleased]: https://github.com/AnderGoig/SwiftInstagram/compare/v1.1.1...develop
[1.1.1]: https://github.com/AnderGoig/SwiftInstagram/tree/v1.1.1
[1.1.0]: https://github.com/AnderGoig/SwiftInstagram/tree/v1.1.0
[1.0.6]: https://github.com/AnderGoig/SwiftInstagram/tree/v1.0.6
[1.0.5]: https://github.com/AnderGoig/SwiftInstagram/tree/v1.0.5
[1.0.4]: https://github.com/AnderGoig/SwiftInstagram/tree/v1.0.4
[1.0.3]: https://github.com/AnderGoig/SwiftInstagram/tree/v1.0.3
[1.0.2]: https://github.com/AnderGoig/SwiftInstagram/tree/v1.0.2
[1.0.1]: https://github.com/AnderGoig/SwiftInstagram/tree/v1.0.1
[1.0.0]: https://github.com/AnderGoig/SwiftInstagram/tree/v1.0.0


Sources/Instagram.swift
//
//  Instagram.swift
//  SwiftInstagram
//
//  Created by Ander Goig on 15/9/17.
//  Copyright © 2017 Ander Goig. All rights reserved.
//

import UIKit

/// A set of helper functions to make the Instagram API easier to use.
public class Instagram {

    // MARK: - Types

    /// Empty success handler
    public typealias EmptySuccessHandler = () -> Void

    /// Success handler
    public typealias SuccessHandler<T> = (_ data: T) -> Void

    /// Failure handler
    public typealias FailureHandler = (_ error: Error) -> Void

    private enum API {
        static let authURL = "https://api.instagram.com/oauth/authorize"
        static let baseURL = "https://api.instagram.com/v1"
    }

    private enum Keychain {
        static let accessTokenKey = "AccessToken"
    }

    enum HTTPMethod: String {
        case get = "GET", post = "POST", delete = "DELETE"
    }

    // MARK: - Properties

    private let urlSession = URLSession(configuration: .default)
    private let keychain = KeychainSwift(keyPrefix: "SwiftInstagram_")

    private var client: (id: String?, redirectURI: String?)?

    // MARK: - Initializers

    /// Returns a shared instance of Instagram.
    public static let shared = Instagram()

    private init() {
        if let path = Bundle.main.path(forResource: "Info", ofType: "plist"), let dict = NSDictionary(contentsOfFile: path) as? [String: AnyObject] {
            let clientId = dict["InstagramClientId"] as? String
            let redirectURI = dict["InstagramRedirectURI"] as? String
            client = (clientId, redirectURI)
        }
    }

    // MARK: - Authentication

    /// Starts an authentication process.
    ///
    /// Shows a custom `UIViewController` with Intagram's login page.
    ///
    /// - parameter controller: The `UINavigationController` from which the `InstagramLoginViewController` will be showed.
    /// - parameter scopes: The scope of the access you are requesting from the user. Basic access by default.
    /// - parameter success: The callback called after a correct login.
    /// - parameter failure: The callback called after an incorrect login.
    public func login(from controller: UINavigationController,
                      withScopes scopes: [InstagramScope] = [.basic],
                      success: EmptySuccessHandler?,
                      failure: FailureHandler?) {

        guard client != nil else { failure?(InstagramError.missingClientIdOrRedirectURI); return }

        let authURL = buildAuthURL(scopes: scopes)

        let vc = InstagramLoginViewController(authURL: authURL, success: { accessToken in
            guard self.storeAccessToken(accessToken) else {
                failure?(InstagramError.keychainError(code: self.keychain.lastResultCode))
                return
            }

            controller.popViewController(animated: true)
            success?()
        }, failure: failure)

        controller.show(vc, sender: nil)
    }

    private func buildAuthURL(scopes: [InstagramScope]) -> URL {
        var components = URLComponents(string: API.authURL)!

        components.queryItems = [
            URLQueryItem(name: "client_id", value: client!.id),
            URLQueryItem(name: "redirect_uri", value: client!.redirectURI),
            URLQueryItem(name: "response_type", value: "token"),
            URLQueryItem(name: "scope", value: scopes.joined(separator: "+"))
        ]

        return components.url!
    }

    /// Ends the current session.
    ///
    /// - returns: True if the user was successfully logged out, false otherwise.
    @discardableResult
    public func logout() -> Bool {
        return deleteAccessToken()
    }

    /// Returns whether a user is currently authenticated or not.
    public var isAuthenticated: Bool {
        return retrieveAccessToken() != nil
    }

    // MARK: - Access Token

    /// Store your own authenticated access token so you don't have to use the included login authentication.
    public func storeAccessToken(_ accessToken: String) -> Bool {
        return keychain.set(accessToken, forKey: Keychain.accessTokenKey)
    }

    /// Returns the current access token.
    public func retrieveAccessToken() -> String? {
        return keychain.get(Keychain.accessTokenKey)
    }

    private func deleteAccessToken() -> Bool {
        return keychain.delete(Keychain.accessTokenKey)
    }

    // MARK: -

    func request<T: Decodable>(_ endpoint: String,
                               method: HTTPMethod = .get,
                               parameters: Parameters = [:],
                               success: ((_ data: T?) -> Void)?,
                               failure: FailureHandler?) {

        let urlRequest = buildURLRequest(endpoint, method: method, parameters: parameters)

        urlSession.dataTask(with: urlRequest) { (data, _, error) in
            if let data = data {
                DispatchQueue.global(qos: .utility).async {
                    do {
                        let jsonDecoder = JSONDecoder()
                        jsonDecoder.keyDecodingStrategy = .convertFromSnakeCase
                        let object = try jsonDecoder.decode(InstagramResponse<T>.self, from: data)

                        if let data = object.data {
                            DispatchQueue.main.async {
                                success?(data)
                            }
                        } else if let message = object.meta.errorMessage {
                            DispatchQueue.main.async {
                                failure?(InstagramError.invalidRequest(message: message))
                            }
                        } else {
                            DispatchQueue.main.async {
                                success?(nil)
                            }
                        }
                    } catch let error {
                        DispatchQueue.main.async {
                            failure?(InstagramError.decoding(message: error.localizedDescription))
                        }
                    }
                }
            } else if let error = error {
                failure?(error)
            }
        }.resume()
    }

    private func buildURLRequest(_ endpoint: String, method: HTTPMethod, parameters: Parameters) -> URLRequest {
        let url = URL(string: API.baseURL + endpoint)!.appendingQueryParameters(["access_token": retrieveAccessToken() ?? ""])

        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = method.rawValue

        switch method {
        case .get, .delete:
            urlRequest.url?.appendQueryParameters(parameters)
        case .post:
            urlRequest.httpBody = Data(parameters: parameters)
        }

        return urlRequest
    }
}
